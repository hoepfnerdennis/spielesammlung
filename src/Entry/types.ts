export type EntryProps = {
  name: string;
  description: string;
  players: {
    from: string;
    to: string;
  };
  age: string;
  duration: string;
};
