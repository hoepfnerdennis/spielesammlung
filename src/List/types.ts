interface IBaseGame {
  name: string;
  description: string;
  playersFrom: number;
  playersTo: number;
  age: string;
  duration: string;
  favorite: boolean;
  simpleRules: boolean;
}

export interface IGame extends IBaseGame {
  image: string;
}

enum LinkTypes {
  'Asset',
  'Space',
  'Environment',
  'ContentType',
}

enum SysTypes {
  'Link',
  'Asset',
}

interface IAPIGame extends IBaseGame {
  image?: {
    sys: {
      type: string;
      linkType: LinkTypes;
      id: string;
    };
  };
}

export interface IAsset {
  sys: {
    space: {
      sys: {
        type: string;
        linkType: LinkTypes;
        id: string;
      };
    };
    id: string;
    type: SysTypes;
    createdAt: string;
    updatedAt: string;
    environment: {
      sys: {
        id: string;
        type: SysTypes;
        linkType: LinkTypes;
      };
    };
    revision: number;
    locale: string;
  };
  fields: {
    title: string;
    description: string;
    file: {
      url: string;
      details: {
        size: number;
        image: {
          width: number;
          height: number;
        };
      };
      fileName: string;
      contentType: string;
    };
  };
}

export interface IAPIResponse {
  sys: {
    type: string;
  };
  total: number;
  skip: number;
  limit: number;
  items: [
    {
      sys: {
        space: {
          sys: {
            type: string;
            linkType: LinkTypes;
            id: string;
          };
        };
        id: string;
        type: string;
        createdAt: string;
        updatedAt: string;
        environment: {
          sys: {
            id: string;
            type: string;
            linkType: LinkTypes;
          };
        };
        revision: number;
        contentType: {
          sys: {
            type: string;
            linkType: LinkTypes;
            id: string;
          };
        };
        locale: string;
      };
      fields: IAPIGame;
    }
  ];
  includes?: {
    Asset: IAsset[];
  };
}

/*
{
  "sys": {
    "type": "Array"
  },
  "total": 3,
  "skip": 0,
  "limit": 100,
  
  "includes": {
    "Asset": [
      {
        "sys": {
          "space": {
            "sys": {
              "type": "Link",
              "linkType": "Space",
              "id": "9sxha2f3gm24"
            }
          },
          "id": "2Txvyka47nhfVoh2Sc1OUh",
          "type": "Asset",
          "createdAt": "2020-03-20T14:04:02.201Z",
          "updatedAt": "2020-03-20T14:04:02.201Z",
          "environment": {
            "sys": {
              "id": "master",
              "type": "Link",
              "linkType": "Environment"
            }
          },
          "revision": 1,
          "locale": "en-US"
        },
        "fields": {
          "title": "Mensch ärgere dich nicht",
          "description": "Mensch ärgere dich nicht",
          "file": {
            "url": "//images.ctfassets.net/9sxha2f3gm24/2Txvyka47nhfVoh2Sc1OUh/d277e6dc09f15b5bb5a134ff7082fe4a/photo-1582921017967-79d1cb6702ee",
            "details": {
              "size": 192767,
              "image": {
                "width": 2100,
                "height": 1400
              }
            },
            "fileName": "photo-1582921017967-79d1cb6702ee",
            "contentType": "image/jpeg"
          }
        }
      }
    ]
  }
}
 */
