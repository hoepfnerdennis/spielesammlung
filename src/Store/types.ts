interface IBaseGame {
  name: string;
  description: string;
  playersFrom: number;
  playersTo: number;
  age: string;
  duration: string;
  favorite: boolean;
  simpleRules: boolean;
}

export interface IGame extends IBaseGame {
  image: string;
}

enum LinkTypes {
  'Asset',
  'Space',
  'Environment',
  'ContentType',
}

enum SysTypes {
  'Link',
  'Asset',
}

interface IAPIGame extends IBaseGame {
  image?: {
    sys: {
      type: string;
      linkType: LinkTypes;
      id: string;
    };
  };
}

export interface IAsset {
  sys: {
    space: {
      sys: {
        type: string;
        linkType: LinkTypes;
        id: string;
      };
    };
    id: string;
    type: SysTypes;
    createdAt: string;
    updatedAt: string;
    environment: {
      sys: {
        id: string;
        type: SysTypes;
        linkType: LinkTypes;
      };
    };
    revision: number;
    locale: string;
  };
  fields: {
    title: string;
    description: string;
    file: {
      url: string;
      details: {
        size: number;
        image: {
          width: number;
          height: number;
        };
      };
      fileName: string;
      contentType: string;
    };
  };
}

export interface IAPIResponse {
  sys: {
    type: string;
  };
  total: number;
  skip: number;
  limit: number;
  items: [
    {
      sys: {
        space: {
          sys: {
            type: string;
            linkType: LinkTypes;
            id: string;
          };
        };
        id: string;
        type: string;
        createdAt: string;
        updatedAt: string;
        environment: {
          sys: {
            id: string;
            type: string;
            linkType: LinkTypes;
          };
        };
        revision: number;
        contentType: {
          sys: {
            type: string;
            linkType: LinkTypes;
            id: string;
          };
        };
        locale: string;
      };
      fields: IAPIGame;
    }
  ];
  includes?: {
    Asset: IAsset[];
  };
}

export interface IFilters {
  playersFrom: number | undefined;
  playersTo: number | undefined;
  favorite: boolean | undefined;
  search: string | undefined;
}

export interface IState {
  games: Array<IGame>;
  filters: IFilters;
}

export type Reducer = (state: IState, action: IAction) => IState;

export type Dispatch = React.Dispatch<IAction>;

export interface IAction {
  type: string;
  payload: any;
}
