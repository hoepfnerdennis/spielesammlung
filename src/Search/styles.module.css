.container {
    margin-right: .3rem;
    display: flex;
    align-items: center;
    width: 100%;
}

.input {
    flex: 1;
    height: 1.5rem;
    font-size: 1.1rem;
    border: 2px solid white;
    border-radius: 5px;
    margin-right: .3rem;
}