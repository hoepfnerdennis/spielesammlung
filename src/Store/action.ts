import { useCallback, useState, useEffect } from 'react';
import {
  IGame,
  IAPIResponse,
  FiltersConfig,
  FilterKey,
  FilterValue,
  SetFilterFunc,
  ActiveFiltersMap,
} from './types';
import { mapResultsToGames } from '../utils';

const SPACE_ID = '9sxha2f3gm24';
const API_TOKEN = '7LDIC95TsrYOfZwEQnbAuMHtij97kfk5r1dIRiGqT8M';
// For local testing replace BASE_URI with
// const BASE_URI = 'http://localhost:3001/response.json';
const BASE_URI = `https://cdn.contentful.com/spaces/${SPACE_ID}/environments/master/entries?access_token=${API_TOKEN}`;

const filtersConfig: FiltersConfig = {
  playersFrom: '[lte]',
  playersTo: '[gte]',
  name: '[match]',
  favorite: '',
};

const fetchGamesAction = async (filters: Map<FilterKey, FilterValue>): Promise<IGame[]> => {
  let url = `${BASE_URI}&content_type=game`;
  filters.forEach((value, key) => {
    url += `&fields.${FilterKey[key]}${filtersConfig[FilterKey[key]]}=${value}`; // &fields.playersTo[gte]=${from}`;
  });

  try {
    const response = await fetch(url);
    const data: IAPIResponse = await response.json();
    const gamesFromAPI: IGame[] = mapResultsToGames(data);
    return gamesFromAPI;
  } catch {
    return [];
  }
};

const useData = (): {
  games: IGame[];
  setFilter: SetFilterFunc;
  activeFilters: ActiveFiltersMap;
} => {
  const [games, setGames] = useState<IGame[]>([]);
  const [activeFilters, setActiveFilters] = useState<ActiveFiltersMap>(new Map());

  useEffect(() => {
    const fetchGames = async (): Promise<void> => {
      const newGames = await fetchGamesAction(activeFilters);
      setGames(newGames);
    };
    fetchGames();
  }, [activeFilters]);

  const addFilter = useCallback((key: FilterKey, value: FilterValue): void => {
    setActiveFilters(activeFiltersMap => {
      if (activeFiltersMap.get(key) === value) {
        return activeFiltersMap;
      }
      const newMap = new Map(activeFiltersMap);
      newMap.set(key, value);
      return newMap;
    });
  }, []);

  const removeFilter = useCallback((key: FilterKey): void => {
    setActiveFilters(activeFiltersMap => {
      if (!activeFiltersMap.has(key)) {
        return activeFiltersMap;
      }
      const newMap = new Map(activeFiltersMap);
      newMap.delete(key);
      return newMap;
    });
  }, []);

  const setFilter = useCallback(
    (key: FilterKey) => (value: FilterValue | undefined): void => {
      if (value) {
        addFilter(key, value);
      } else {
        removeFilter(key);
      }
    },
    [addFilter, removeFilter]
  );

  return { games, setFilter, activeFilters };
};

export default useData;
