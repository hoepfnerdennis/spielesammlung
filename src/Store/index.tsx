import React, { createContext, useReducer } from 'react';
import { IState, Reducer, IAction } from './types';

export const FETCH_DATA_FINISHED = 'FETCH_DATA_FINISHED';
export const SET_SEARCH_TERM = 'SET_SEARCH_TERM';
export const SET_PALYER_FROM = 'SET_PALYER_FROM';
export const SET_PLAYER_TO = 'SET_PLAYER_TO';
export const SET_FAVORITE = 'SET_FAVORITE';

const initialState: IState = {
  games: [],
  filters: {
    playersFrom: undefined,
    playersTo: undefined,
    favorite: undefined,
    search: undefined,
  },
};

export const Store = createContext<IState | any>(initialState);

const reducer = (state: IState, action: IAction): IState => {
  switch (action.type) {
    case FETCH_DATA_FINISHED:
      return { ...state, games: action.payload };

    case SET_PALYER_FROM:
      return { ...state, filters: { ...state.filters, playersFrom: action.payload } };

    case SET_PLAYER_TO:
      return { ...state, filters: { ...state.filters, playersTo: action.payload } };

    case SET_SEARCH_TERM:
      return { ...state, filters: { ...state.filters, search: action.payload } };

    case SET_FAVORITE:
      return { ...state, filters: { ...state.filters, favorite: action.payload } };

    default:
      return state;
  }
};

export const StoreProvider: React.SFC = ({ children }): JSX.Element => {
  const [state, dispatch] = useReducer<Reducer>(reducer, initialState);
  return <Store.Provider value={{ state, dispatch }}>{children}</Store.Provider>;
};
