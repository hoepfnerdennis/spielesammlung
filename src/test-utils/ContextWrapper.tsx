import React, { FunctionComponent } from 'react';
import { createRenderer } from 'fela';
import { RendererProvider } from 'react-fela';

const renderer = createRenderer();

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const ContextWrapper: FunctionComponent = ({ children }) => {
  return <RendererProvider renderer={renderer}>{children}</RendererProvider>;
};

export default ContextWrapper;
