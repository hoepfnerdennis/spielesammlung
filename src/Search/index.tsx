import React, { useState } from 'react';
import styles from './styles.module.css';
import { FilterKey } from '../Store/types';
import Button from '../ Button';
import { useSetFilter } from '../Store/action';

const Search: React.FC = (): JSX.Element => {
  const [searchTerm, setSearchTermInternal] = useState<string>('');
  const setFilter = useSetFilter();

  const doSearch = (): void => {
    setFilter(FilterKey.name)(searchTerm);
  };

  const handleSubmit = (event: React.FormEvent<HTMLFormElement>): void => {
    event.preventDefault();
    doSearch();
  };

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>): void => {
    setSearchTermInternal(event.target.value);
  };

  return (
    <form onSubmit={handleSubmit} className={styles.container}>
      <input
        type="text"
        value={searchTerm}
        onChange={handleChange}
        className={styles.input}
        placeholder="Spiel suchen..."
      />
      <Button data-testid="submit" onClick={doSearch} secondary>
        Suchen
      </Button>
    </form>
  );
};

export default Search;
