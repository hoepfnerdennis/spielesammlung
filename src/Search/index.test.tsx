import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Search from './index';
import { FilterKey } from '../Store/types';
import ContextWrapper from '../test-utils/ContextWrapper';
import { useSetFilter } from '../Store/action';

jest.mock('../Store/action');

// @ts-ignore
// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
jest.mock('../Button', () => ({ secondary, onClick, children }) => (
  <button type="button" data-checked={secondary} onClick={onClick}>
    {children}
  </button>
));

describe('Search', () => {
  it('should render and match snapshot', () => {
    const { container } = render(<Search />);
    expect(container).toMatchSnapshot();
  });
  it('should handle input', () => {
    const setSearchTerm = jest.fn();
    const setFilter = jest.fn(() => setSearchTerm);
    // @ts-ignore
    useSetFilter.mockImplementation(() => setFilter);
    const { getByPlaceholderText, getByText } = render(
      <ContextWrapper>
        <Search />
      </ContextWrapper>
    );
    fireEvent.change(getByPlaceholderText('Spiel suchen...'), { target: { value: 'abc' } });
    fireEvent.submit(getByText('Suchen'));
    expect(setFilter).toHaveBeenCalledWith(FilterKey.name);
    expect(setSearchTerm).toHaveBeenCalledWith('abc');
  });
});
