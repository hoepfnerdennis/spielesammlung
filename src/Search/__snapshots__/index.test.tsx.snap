// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search should render and match snapshot 1`] = `
<div>
  <form
    class="container"
  >
    <input
      class="input"
      placeholder="Spiel suchen..."
      type="text"
      value=""
    />
    <button
      data-checked="true"
      type="button"
    >
      Suchen
    </button>
  </form>
</div>
`;
